idea/src/org/jetbrains/kotlin/idea/inspections/RedundantAsDynamicInspection.kt

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.inspections
+
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.openapi.project.Project
+import org.jetbrains.kotlin.idea.intentions.getCallableDescriptor
+import org.jetbrains.kotlin.idea.project.platform
+import org.jetbrains.kotlin.js.resolve.JsPlatform
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.callExpressionVisitor
+import org.jetbrains.kotlin.psi.psiUtil.getQualifiedExpressionForSelector
+import org.jetbrains.kotlin.types.DynamicType
+
+class RedundantAsDynamicInspection : AbstractKotlinInspection() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) =
+        callExpressionVisitor(fun(call) {
+            if (call.platform != JsPlatform) return
+            if (call.calleeExpression?.text != "asDynamic") return
+            if (call.getQualifiedExpressionForSelector()?.receiverExpression?.getCallableDescriptor()?.returnType !is DynamicType) return
+            holder.registerProblem(
+                call,
+                "Redundant 'asDynamic' call",
```

Comment:
It seems to me that the message is a bit misleading. The important part here is that the code might fail at run time, because the programmer doesn't realize the `asDynamic` is a member, not the extension function from stdlib s/he might be thinking of.

So maybe it should be something along the lines of "Suspicious 'asDynamic' member invocation" ? "Possibly eroneous"? Something else?

In any case I don't think the word "redundant" conveys the actual problem properly.